### DISCUSSION TITLE
Docker pretix upgrade issue

### DISCUSSION THREAD

**Discussion by jxs714** [pretix/pretix#4045] [Q&A] [answered]
Hey guys, 

Forgot to mention this a while back. The docker command for pretix upgrade does not work in the documentation of pretix. This is the command I am using:

docker exec -it CONTAINERNAME pretix upgrade. The results are below:

Operations to perform:
  Apply all migrations: auth, badges, banktransfer, contenttypes, oauth2_provider, otp_static, otp_totp, paypal, pretix_bounces, pretix_pages, pretixapi, pretixbase, pretixdroid, pretixhelpers, pretixmultidomain, sendmail, sessions, stripe, ticketoutputpdf
Running migrations:
  No migrations to apply.

The only way I can upgrade pretix docker container is by doing 1 of the 3 commands for manual deployment below:

**OPTION 1 - Update in Docker (NON Root)**
docker exec -it CONTAINERNAME pip install --upgrade pretix
Reboot docker container.

The error I received before docker container reboot was this:
WARNING: The script markdown_py is installed in '/pretix/.local/bin' which is not on PATH.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.

**OPTION 2 - Update in Docker as Root**
docker exec -it -u root CONTAINERNAME sudo pip install --upgrade pretix
Reboot docker container.

The error I received before docker container reboot was this:
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv

When doing either manual update above, pretix either does not detect the latest version as updated for base pretix and plugins when checking for updates in the backend admin dashboard. However, it shows the updated version number when running "pip show pretix" in command line inside the docker such as this

docker exec -it CONTAINERNAME pip show pretix

Two questions:
1. Both of these update steps seems to have worked. Should these errors be ignored?
2. Are these proper alternative upgrade procedures and/or are there any changes that need to be done in these commands? What is the reason why the pretix dashboard does not show the latest version updated even though it shows its update in command line of the docker container?

**OPTION 3**
Installing a new pretix container and re-installing all plugins again.

Please advise, thanks

**Comment by jxs714** [comment]
Hey guys,

Any updates on this question? I see updates on Pretix and would like to safely upgrade to be up to date.

**Reply by raphaelm** [reply]
Both of the options you present are wrong and go against how containers are supposed to work. Please never run pip inside a container.

If installed according to the docs, the three commands here do work:
https://docs.pretix.eu/en/latest/admin/installation/docker_smallscale.html#updates

But of course only if you use all of them.

**Reply by jxs714** [reply]
Hi Raphael,

Hope all has been well. Thanks for the update. Just followed the instructions again and did the upgrade, and although it is now up to date, the new container does not have the plugins I've installed in the previous container. The upgrade process is easy to do and wasn't the issue, the main issue was being able to keep and or update the plugins as well. 

How do I commit to the upgrade while keeping the plugins? Is there a way to do this in the same way? or do I have to upgrade pretix as I did and rebuild the image with plugins by re-running the dockerfile plugin install script?

**Reply by raphaelm** [reply, marked as answer]
Yes, you have to recreate the image with your plugins based on the new pretix version.

**Reply by jxs714** [reply]
Hi Raphael,

Thanks for the update and response. No problem!
